### ISSUE TITLE
Translation errors

### ISSUE THREAD

**Issue by AA-Turner** [sphinx-doc/sphinx-doc-translations#36]
I've had to blacklist Tamil, and zh_CN is newly failing, meaning the translations PRs don't get updated.

https://github.com/sphinx-doc/sphinx/actions/workflows/transifex.yml

Is there a way of ensuring in transifex that these errors don't happen?

A

**Comment by rffontenelle** [comment]
These errors are for UI, not docs. Considering reporting in there. 

Important subject, though.

**Comment by rffontenelle** [comment]
@AA-Turner I already mentioned the Tamil issue in #28

**Comment by rffontenelle** [comment]
> Is there a way of ensuring in transifex that these errors don't happen?

@AA-Turner Sorry for not answering this question specific before. 

TL;DR; I don't think there is straightforward way to avoid it in Transifex.

Translations not honoring placeholders `%s` are reported in-screen as error for the translations (see [Tamil example](https://app.transifex.com/sphinx-doc/sphinx-1/translate/#ta/$?q=check%3Aerror)), but I haven't found a straightforward way to `tx pull` filtering these error strings. Haven't found a [API](https://developers.transifex.com/reference) endpoint either.

Pulling only reviewed translations would reduce the chance of these errors, but not ensure (plus adding a big burden to the existing contributors). I don't think it is worth.

A manual solution would be to have me editing and fixing, or clearing the problematic translation strings. I can do that if you need, but I need to be made aware (via CI etc.) whenever it happens.

Is it possible to programmatically retrieve the language codes causing the compilation to fail?  It occurred to me that the transifex.yml CI workflow could keep going by first clearing these problematic language codes with `git checkout <lang>`.

**Comment by n-peugnet** [comment]
> A manual solution would be to have me editing and fixing, or clearing the problematic translation strings. I can do that if you need, but I need to be made aware (via CI etc.) whenever it happens.

I just thought about it, but maybe the simplest solution would be to mark all failing messages as _fuzzy_. I didn't try but it should skip these messages, allowing the rest to compile without errors.

> Is it possible to programmatically retrieve the language codes causing the compilation to fail?

It is possible with [`msgfmt --check` (from gettext)](https://www.gnu.org/software/gettext/manual/html_node/msgfmt-Invocation.html#index-_002dc_002c-msgfmt-option). I made a [pull request on Sphinx](https://github.com/sphinx-doc/sphinx/pull/12394) to do exactly this for the internal messages.

Maybe another script could use this information to add the _fuzzy_ tag automatically.

**Comment by n-peugnet** [comment]
A probably simpler possibility would be to make a `babel_runner.py check` command that would only return the errors found by Babel. This way there is no dependency on gettext and the python script I showed earlier could be used to implement some kind of `babel_runner.py check --fix` command.
The only inconvenient is that we can now miss the errors discovered by `msgfmt --check`, but nothing prevent us from adding them back in the python script later.

**Comment by rffontenelle** [comment]
Just to mention that Tamil team fixed the errors reported by Transifex, although I haven't checked the quality of the rest of the docs. Last time I checked, Changelog translation didn't have the issue number and link, so there's still room for improvement.
